### PR TITLE
`ios_base` constants should have bitmask types

### LINKED ISSUE #3401: <ios>: std::ios_base::openmode is not a bitmask type
**Describe the bug**
std::ios_base::openmode is not implemented as a bitmask type, therefore, for example, |= operator usage fails to compile.

**Command-line test case**
```
C:\Temp>type example.cpp
#include <ios>

int main()
{
    auto open_flags = std::ios::binary;
    open_flags |= std::ios::app;

    return 0;
}

C:\Temp>cl -std:c++17 .\example.cpp

example.cpp
<source>(6): error C2678: binary '|=': no operator found which takes a left-hand operand of type 'std::_Iosb<int>::_Openmode' (or there is no acceptable conversion)
C:/data/msvc/14.21.27702.2/include\cstddef(70): note: could be 'std::byte &std::operator |=(std::byte &,const std::byte) noexcept' [found using argument-dependent lookup]
<source>(6): note: while trying to match the argument list '(std::_Iosb<int>::_Openmode, const std::_Iosb<int>::_Openmode)'
Compiler returned: 2

```

**Expected behavior**
compile successfully 

**STL version**
    ```
    Microsoft (R) C/C++ Optimizing Compiler Version 19.21.27702.2 for x64
    ```

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #3401 

If I understand correctly we can't change types of the variables but we can add operators.

### CODE REVIEW

**Comment by AlexGuteniev** [comment]
Have you considered `_BITMASK_OPS` macro?

**Comment by fsb4000** [comment]
Thanks. I knew about the macro but I forgot about it.

**Comment by frederick-vs-ja** [comment]
`std::ios_base::openmode` is still `int`. We should fix the type mismatch, see [[ios.base.general]](https://eel.is/c++draft/ios.base.general).

**Comment by AlexGuteniev** [comment]
> `std::ios_base::openmode` is still `int`. We should fix the type mismatch, see [[ios.base.general]](https://eel.is/c++draft/ios.base.general).

Not sure if it wouldn't be an ABI break

**Comment by fsb4000** [comment]
Can we fix it?

What if a user used `std::ios_base::openmode` in his function. Before the type was `int`,  and it is `std::ios_base::_Openmode` after.

It looks like ABI break.

I don't really understand all this about ABI break, but I remember that `intmax_t` was considered a very bad decision because this type is forever `int64_t`, because changing it to a hypothetical `int128_t` would be an ABI break.

**Comment by StephanTLavavej** [comment]
The variables don't seem to appear on the DLL's export surface, so it might be possible to change their types. Yes, theoretically someone could have been depending on their types in their own signatures, but it would be fairly hard (especially pre-`decltype`). I am pretty scared to change anything in iostreams but it might be possible, and it would be simpler and better for throughput than emitting a bunch of bitmask ops (which are unfortunately widely visible).

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: tests/std/tests/P2467R1_exclusive_mode_fstreams/test.cpp [hunk 1]
```diff
@@ -44,6 +45,81 @@ void test_file_create_fail(const std::ios_base::openmode bad_mode) {
...
+    static_assert(is_same_v<remove_const_t<decltype(IB::right)>, IB::fmtflags>);
+    static_assert(is_same_v<remove_const_t<decltype(IB::internal)>, IB::fmtflags>);
+    static_assert(is_same_v<remove_const_t<decltype(IB::dec)>, IB::fmtflags>);
+    static_assert(is_same_v<remove_const_t<decltype(IB::oct)>, IB::fmtflags>);
+    static_assert(is_same_v<remove_const_t<decltype(IB::hex)>, IB::fmtflags>);
+    static_assert(is_same_v<remove_const_t<decltype(IB::scientific)>, IB::fmtflags>);
+    static_assert(is_same_v<remove_const_t<decltype(IB::fixed)>, IB::fmtflags>);
+    static_assert(is_same_v<remove_const_t<decltype(IB::boolalpha)>, IB::fmtflags>);
+    static_assert(is_same_v<remove_const_t<decltype(IB::adjustfield)>, IB::fmtflags>);
+    static_assert(is_same_v<remove_const_t<decltype(IB::boolalpha)>, IB::fmtflags>);
```
Comment (StephanTLavavej) [inline tests/std/tests/P2467R1_exclusive_mode_fstreams/test.cpp:116]: Repeated `boolalpha`.

File: tests/std/tests/P2467R1_exclusive_mode_fstreams/test.cpp [hunk 2]
```diff
@@ -44,6 +45,81 @@ void test_file_create_fail(const std::ios_base::openmode bad_mode) {
     assert(!fs::exists(test_file));
 }
 
+// Also test GH-3401: <ios>: std::ios_base::openmode is not a bitmask type
+constexpr bool test_gh_3401() {
+    using IB = std::ios_base;
+    using namespace std;
```
Comment (StephanTLavavej) [inline tests/std/tests/P2467R1_exclusive_mode_fstreams/test.cpp:51]: Nitpick: These could be swapped.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks, I pushed small changes to the test.

I double-checked for changes to the release and debug DLL export surfaces, and found none. I think this is ABI-safe from a "does this affect the redist" perspective, and I think that it's unlikely to be disruptive (as I mentioned earlier, *theoretically* someone could have used `decltype` to make their own signatures depend on the types of these constants, but that is totally different from normal usage).

There is now unused enum machinery but I would prefer not to mess with that at this time (as the TRANSITION, ABI comments indicate, I once thought this affected ABI - I can't remember why).

**Review by strega-nil-ms — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for fixing this long-standing nonconformance! :sparkles: :smiley_cat: :tada:

**Pull request merged**